### PR TITLE
GR-990 Add namespace handling to client

### PR DESCRIPTION
**Pull request by qsun0621**

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -64,3 +64,12 @@ Close client connection
 func (c *Client) Close() {
 	closeChannel(&c.Channel, &c.methods)
 }
+
+func (c *Client) ConnectNamespace(namespace string) error {
```
Comment (jgiles) [inline client.go:73]: Can we add a comment explaining what this is doing?

Is it setting the default namespace for the underlying connection? Or does it just make this connection eligible to call `Emit` with this namespace in the future?

File: protocol/socketio.go [hunk 2]
```diff
@@ -50,6 +50,8 @@ func Encode(msg *Message) (string, error) {
 		return "", err
 	}
 
+	result = result + msg.Namespace
```
Comment (jgiles) [inline protocol/socketio.go:53]: Namespace is just appended to the msg type string?

Are there any protocol reference docs you could tag here?

File: protocol/socketio.go [hunk 3]
```diff
@@ -64,15 +66,15 @@ func Encode(msg *Message) (string, error) {
 	}
 
 	if msg.Type == MessageTypeAckResponse {
-		return result + "[" + msg.Args + "]", nil
+		return result + ",[" + msg.Args + "]", nil
```
Comment (jgiles) [inline protocol/socketio.go:69]: similarly, would be good to have a protocol reference here.

File: send.go [hunk 4]
```diff
@@ -50,10 +50,11 @@ func send(msg *protocol.Message, c *Channel, args interface{}) error {
 /**
 Create packet based on given data and send it
 */
-func (c *Channel) Emit(method string, args interface{}) error {
+func (c *Channel) Emit(method string, namespace string, args interface{}) error {
```
Comment (jgiles) [inline send.go:53]: This sends in a specific namespace? Does one have to connect to that same namespace previously?

File: protocol/socketio.go (hunk 2 above)
Comment (qsun0621) [reply protocol/socketio.go:53]: Nope. I guessed by printing out the message sent by Python socketio client.

File: send.go (hunk 4 above)
Comment (qsun0621) [reply send.go:53]: Yes. You need to establish the connection to that namespace before you can start emitting message in it.

File: protocol/socketio.go (hunk 3 above)
Comment (qsun0621) [reply protocol/socketio.go:69]: Could not find any specific protocol doc. Also guessed from the message logged by Python socketio client.

File: client.go (hunk 1 above)
Comment (qsun0621) [reply client.go:73]: Default is no namespace. It allows multiple namespaces to be connectioned in the same session. And you do need to connect to one specific namespace before sending request to that particular namespace.
Comments added.

File: protocol/socketio.go (hunk 2 above)
Comment (jgiles) [reply protocol/socketio.go:53]: 😂

**Review by jgiles — APPROVED** [review]

**Pull request merged**